File: recursos/codigo/vulneravel.py
"""
=========================================================================
  ⚠️  CÓDIGO DELIBERADAMENTE VULNERÁVEL — MATERIAL DIDÁTICO (M11)  ⚠️
=========================================================================

NÃO USE NADA DESTE ARQUIVO EM PRODUÇÃO.

Cada view abaixo contém pelo menos uma vulnerabilidade das classes do
OWASP Top 10:2021. O exercício do M11 é, em duplas:

  1. identificar a vulnerabilidade (e nomeá-la);
  2. demonstrar a exploração com um payload/URL concreto, no seu próprio
     ambiente local;
  3. corrigir;
  4. explicar por que a correção funciona.

O gabarito está no fim do arquivo, comentado. Não leia antes de tentar.

Regra da disciplina: exploração apenas contra o SEU ambiente local ou
contra sistemas para os quais você tem autorização explícita por escrito.
"""

import os
import subprocess

from django import forms
from django.contrib.auth import get_user_model
from django.http import FileResponse, HttpResponse
from django.shortcuts import get_object_or_404, redirect, render
from django.views.decorators.csrf import csrf_exempt

from acervo.models import Emprestimo, Obra

Usuario = get_user_model()


# --------------------------------------------------------------------- 1
def busca(request):
    termo = request.GET.get("q", "")
    obras = Obra.objects.raw(
        f"SELECT * FROM acervo_obra WHERE titulo LIKE '%{termo}%'"
    )
    return render(request, "busca.html", {"obras": obras})


# --------------------------------------------------------------------- 2
def emprestimo_detail(request, pk):
    emprestimo = get_object_or_404(Emprestimo, pk=pk)
    return render(request, "emprestimo_detail.html", {"emprestimo": emprestimo})


# --------------------------------------------------------------------- 3
@csrf_exempt
def excluir_obra(request, pk):
    Obra.objects.filter(pk=pk).delete()
    return redirect("acervo:obra_list")


# --------------------------------------------------------------------- 4
#   template obra_detail.html:
#       <div class="sinopse">{{ obra.sinopse|safe }}</div>
#       <a href="{{ obra.link_externo }}">Saiba mais</a>


# --------------------------------------------------------------------- 5
def download(request):
    nome = request.GET["arquivo"]
    return FileResponse(open(os.path.join("/var/media/", nome), "rb"))


# --------------------------------------------------------------------- 6
def login_view(request):
    if request.method != "POST":
        return render(request, "login.html")
    u = Usuario.objects.filter(username=request.POST["usuario"]).first()
    if not u:
        return render(request, "login.html", {"erro": "Usuário não encontrado"})
    if u.password != request.POST["senha"]:
        return render(request, "login.html", {"erro": "Senha incorreta"})
    request.session["user_id"] = u.id
    return redirect("acervo:home")


# --------------------------------------------------------------------- 7
class ObraForm(forms.ModelForm):
    class Meta:
        model = Obra
        fields = "__all__"


# --------------------------------------------------------------------- 8
def relatorio(request):
    if request.GET.get("admin") == "1":
        return render(request, "relatorio_completo.html", {"dados": "..."})
    return render(request, "relatorio_basico.html")


# --------------------------------------------------------------------- 9
def gerar_miniatura(request):
    arquivo = request.GET["arquivo"]
    os.system(f"convert {arquivo} -resize 200x200 /tmp/thumb.png")
    return FileResponse(open("/tmp/thumb.png", "rb"))


# -------------------------------------------------------------------- 10
def exportar_csv(request):
    linhas = ["id,titulo,associado,email,cpf,telefone"]
    for e in Emprestimo.objects.select_related("associado"):
        a = e.associado
        linhas.append(f"{e.pk},{e.exemplar.obra.titulo},{a.nome},{a.email},{a.cpf},{a.telefone}")
    return HttpResponse("\n".join(linhas), content_type="text/csv")


# =====================================================================
# GABARITO — não leia antes de tentar.
# =====================================================================
#
#  1. A03 Injeção de SQL. Payload: ?q=' OR '1'='1' --
#     Correção: Obra.objects.filter(titulo__icontains=termo)
#     Ou, se raw for indispensável:
#         Obra.objects.raw("SELECT * FROM acervo_obra WHERE titulo LIKE %s",
#                          [f"%{termo}%"])
#
#  2. A01 IDOR + falta de autenticação. Qualquer pessoa acessa
#     /emprestimos/<qualquer-id>/. Correção: exigir login e filtrar o
#     queryset pelo usuário; devolver 404 (não 403) para o que não é seu.
#
#  3. CSRF (A01) + método errado + sem autorização. Exclusão por GET,
#     sem token e sem checagem de permissão. Correção: @require_POST,
#     remover @csrf_exempt, @permission_required, e tela de confirmação.
#
#  4. A03 XSS armazenado. |safe em texto do usuário executa
#     <script>. E href com valor do usuário permite javascript:.
#     Correção: remover |safe (ou sanitizar na escrita com nh3/bleach) e
#     validar o esquema da URL no servidor (apenas http/https).
#
#  5. A01 Path traversal. Payload: ?arquivo=../../etc/passwd
#     Correção: django.utils._os.safe_join, validar o nome contra uma
#     lista/padrão, e servir apenas arquivos referenciados por um model
#     ao qual o usuário tem acesso.
#
#  6. A07 Várias: (a) enumeração de usuários por mensagens distintas;
#     (b) senha comparada em texto puro; (c) comparação não é em tempo
#     constante; (d) sessão manipulada na mão, sem rotação de chave.
#     Correção: authenticate() + login() do Django, mensagem única.
#
#  7. A01 Mass assignment. fields="__all__" expõe qualquer campo que
#     venha a ser adicionado ao model. Correção: lista explícita.
#
#  8. A01 Autorização por parâmetro do cliente. ?admin=1 dá acesso.
#     Correção: decidir pelo request.user e por permissão.
#
#  9. A03 Injeção de comando. Payload: ?arquivo=x.jpg; rm -rf /
#     Correção: subprocess.run([...], check=True) sem shell, com o nome
#     validado; melhor ainda, usar Pillow em vez de shell.
#
# 10. A01/A02 + LGPD. Endpoint sem autenticação expondo dados pessoais
#     (e-mail, CPF, telefone) de todos os associados; CSV sem escape
#     (injeção de fórmula em planilha: campo iniciado por "=" ou "+").
#     Correção: exigir permissão, minimizar colunas, escapar o CSV com o
#     módulo csv, e questionar se o CPF precisa existir no sistema.

File: recursos/codigo/vulneravel.ts
/**
 * M13 — Laboratório de vulnerabilidades (backend).
 *
 * ⚠️ CÓDIGO DELIBERADAMENTE INSEGURO. Existe para ser atacado e corrigido em
 *    sala. Nunca use nada daqui em produção.
 *
 * São 10 falhas. Para cada uma, entregue:
 *   1. o nome OWASP (A01, A03, …)
 *   2. o impacto de negócio, na linguagem da biblioteca
 *   3. a exploração concreta (payload, URL ou comando)
 *   4. a correção
 *   5. por que a correção funciona
 *
 * O gabarito está no fim do arquivo. Não leia antes de tentar.
 */

import { exec } from "node:child_process";
import { join } from "node:path";
import { Body, Controller, Delete, Get, Param, Patch, Post, Query, Res } from "@nestjs/common";

@Controller("inseguro")
export class InseguroController {
  constructor(
    private readonly obras: Repository<Obra>,
    private readonly usuarios: Repository<Usuario>,
    private readonly emprestimos: Repository<Emprestimo>,
  ) {}

  // ---------------------------------------------------------------- (1)
  @Get("buscar")
  async buscar(@Query("q") termo: string) {
    return this.obras
      .createQueryBuilder("obra")
      .where(`obra.titulo LIKE '%${termo}%'`)
      .getMany();
  }

  // ---------------------------------------------------------------- (2)
  @Get("emprestimos/:id")
  async verEmprestimo(@Param("id") id: number) {
    return this.emprestimos.findOne({
      where: { id },
      relations: { associado: true, exemplar: true },
    });
  }

  // ---------------------------------------------------------------- (3)
  @Get("usuarios")
  async listarUsuarios() {
    return this.usuarios.find();
  }

  // ---------------------------------------------------------------- (4)
  @Patch("perfil/:id")
  async atualizarPerfil(@Param("id") id: number, @Body() corpo: any) {
    await this.usuarios.update(id, corpo);
    return this.usuarios.findOneBy({ id });
  }

  // ---------------------------------------------------------------- (5)
  @Delete("obras/:id")
  async removerObra(@Param("id") id: number) {
    await this.obras.delete(id);
    return { removido: true };
  }

  // ---------------------------------------------------------------- (6)
  @Get("capa")
  baixarCapa(@Query("arquivo") arquivo: string, @Res() res: Response) {
    return res.sendFile(join("/var/bibliocom/midia", arquivo));
  }

  // ---------------------------------------------------------------- (7)
  @Post("converter")
  converter(@Body("nome") nome: string) {
    exec(`convert /tmp/${nome} /tmp/saida.png`);
    return { ok: true };
  }

  // ---------------------------------------------------------------- (8)
  @Post("entrar")
  async entrar(@Body() dto: { email: string; senha: string }) {
    const usuario = await this.usuarios.findOne({
      where: { email: dto.email },
      select: { id: true, email: true, senhaHash: true },
    });
    if (!usuario) {
      return { erro: "Não existe conta com este e-mail" };
    }
    if (usuario.senhaHash !== md5(dto.senha)) {
      return { erro: "Senha incorreta" };
    }
    return { id: usuario.id, email: usuario.email };
  }

  // ---------------------------------------------------------------- (9)
  @Get("relatorio")
  async relatorio(@Query("tamanho") tamanho: number) {
    return this.obras.find({ take: tamanho });
  }

  // ---------------------------------------------------------------- (10)
  @Get("erro")
  async comErro(@Query("id") id: number) {
    try {
      return await this.obras.findOneByOrFail({ id });
    } catch (err) {
      return { mensagem: err.message, stack: err.stack, sql: err.query };
    }
  }
}

/*
================================================================================
GABARITO — não leia antes de tentar.
================================================================================

(1) A03 — Injeção de SQL.
    O termo é concatenado na consulta. `?q=%' OR 1=1 --` devolve o acervo
    inteiro; `UNION SELECT` alcança outras tabelas.
    Correção: `.where("obra.titulo LIKE :termo", { termo: `%${termo}%` })`.
    Funciona porque o driver envia comando e parâmetro separados: o banco nunca
    interpreta o dado como SQL.

(2) A01 — IDOR (referência direta insegura).
    Qualquer pessoa lê o empréstimo de qualquer outra trocando o `:id`.
    Correção: filtrar pela sessão dentro da consulta — quem não pode ver recebe
    404 e não descobre nem que o registro existe.

(3) A01/A02 — Exposição de dados sensíveis.
    Devolve a entidade inteira, com `senhaHash` e e-mail de todos. Sem
    autenticação, ainda por cima.
    Correção: guard de papel + DTO de saída que declara o que sai.

(4) A01 — Mass assignment com escalada de privilégio.
    `corpo: any` grava qualquer campo: enviar `{"papel":"coordenacao"}` promove o
    próprio usuário.
    Correção: DTO tipado + `ValidationPipe` com `whitelist: true`, e o papel
    definido no servidor.

(5) A01 — Falta de autorização.
    Exclusão sem checar quem chama. Um `curl` apaga o acervo.
    Correção: `@UseGuards` + `@Papeis(Papel.COORDENACAO)`.

(6) A01 — Path traversal.
    `?arquivo=../../etc/passwd` sai da pasta de mídia.
    Correção: resolver o caminho e conferir o prefixo antes de servir.

(7) A03 — Injeção de comando.
    `nome` vai para o shell: `foo; rm -rf /` executa.
    Correção: `execFile("convert", [entrada, saida])` — argumentos separados,
    sem shell interpretando.

(8) A07 — Três falhas de autenticação numa função só.
    (a) Enumeração de usuários: mensagens diferentes revelam quais e-mails
        existem. Use uma única mensagem genérica.
    (b) MD5: rápido demais, quebrado. Use Argon2 ou bcrypt.
    (c) Comparação com `!==`: vulnerável a ataque de tempo. Use a função de
        verificação da própria biblioteca de hash.

(9) A05 — Negação de serviço por parâmetro sem teto.
    `?tamanho=99999999` consome memória do servidor com uma requisição barata.
    Correção: `@Max(100)` no DTO da query.

(10) A05 — Vazamento de informação pelo erro.
     Devolve `stack` e a consulta SQL ao cliente: revela caminhos de arquivo,
     estrutura do banco e versões.
     Correção: registrar o detalhe no log e devolver mensagem genérica com 500.
     O padrão do NestJS já faz isso — o risco é você "melhorar" a depuração.
================================================================================
*/
